src/tensor_ops/reduce.rs

```diff
@@ -0,0 +1,48 @@
+use crate::prelude::*;
+
+/// TODO
+pub trait Reduce1<const I: isize>: Tensor<Dtype = f32> {
+    type Reduced: Tensor<Dtype = Self::Dtype, Tape = Self::Tape> + Broadcast1<Self, I>;
+    type DeviceR: Reduce1Axis<Self::Array, <Self::Reduced as HasArrayType>::Array, I>;
+}
+
+macro_rules! reduction {
+    ($SrcTy:tt, [$($SrcDims:tt),*], $DstTy:tt, [$($DstDims:tt),*], $Axis:expr) => {
+impl<$(const $SrcDims: usize, )* H: Tape> Reduce1<$Axis> for $SrcTy<$($SrcDims, )* H> {
+    type Reduced = $DstTy<$($DstDims, )* H>;
+    type DeviceR = <Self as HasDevice>::Device;
+}
+    };
+}
+
+// 0d
+// impl<H: Tape> Reduce1<0> for Tensor0D<H> {
+//     type Reduced = Self;
+//     type DeviceR = <Self as HasDevice>::Device;
+// }
+impl<H: Tape> Reduce1<-1> for Tensor0D<H> {
+    type Reduced = Self;
+    type DeviceR = <Self as HasDevice>::Device;
+}
+
+// 1d
+// reduction!(Tensor1D, [M], Tensor0D, [], 0);
+reduction!(Tensor1D, [M], Tensor0D, [], -1);
+
+// 2d
+reduction!(Tensor2D, [M, N], Tensor1D, [N], 0);
+// reduction!(Tensor2D, [M, N], Tensor1D, [M], 1);
+reduction!(Tensor2D, [M, N], Tensor1D, [M], -1);
+
+// 3d
+reduction!(Tensor3D, [M, N, O], Tensor2D, [N, O], 0);
+reduction!(Tensor3D, [M, N, O], Tensor2D, [M, O], 1);
+// reduction!(Tensor3D, [M, N, O], Tensor2D, [M, N], 2);
+reduction!(Tensor3D, [M, N, O], Tensor2D, [M, N], -1);
```

Comment:
Having multiple impls for the same axis (i.e. the last axis can be 2 or -1), means the compiler complains about multiple trait impls that require explicitly setting that.

I.e. with both -1 and 2 available for reducing/broadcasting:
```rust
let a: Tensor2D<3, 5> = ...;
let b: Tensor3D<3, 5, 7> = a.broadcast1(); // compiler will complain about multiple impls here
let c= Broadcast1<Tensor3D<3, 5, 7>, -1>::broadcast1(b); // this will be fine
```

However if just -1 is available (or 2) then you can use the `.broadcast1()`.